### PR TITLE
fix(reverse-proxy): allow multi-label subdomains in service input

### LINKED ISSUE #667: Reverse Proxy service modal strips dots from the subdomain input, blocking nested subdomains the API accepts
### Describe the problem

When creating a Reverse Proxy **service**, the subdomain input in the dashboard silently strips any character that is not `[a-z0-9-]`, including dots (`.`). This makes it impossible to create a service on a **nested (multi-label) subdomain** such as `dev.app.example.com` — typing the dot just deletes it, so you can only ever enter a single label (`app.example.com`).

The management API, however, accepts the full multi-label domain and provisions everything correctly: the ACME (TLS-ALPN-01) certificate is issued, the service becomes `active`, and requests are served / redirected to SSO as expected. So the UI is stricter than the backend, with no functional reason.

### Root cause

`src/modules/reverse-proxy/domain/ReverseProxyDomainInput.tsx` — the subdomain field sanitizes input on change:

```tsx
onChange={(e) =>
  onSubdomainChange(e.target.value.toLowerCase().replace(/[^a-z0-9-]/g, ""))
}
```

`/[^a-z0-9-]/g` removes `.`, so nested subdomains can never be entered. A single DNS label is `[a-z0-9-]`, but a *subdomain* may legitimately contain multiple dot-separated labels.

### To reproduce

1. Self-hosted NetBird with the Reverse Proxy feature and a connected proxy cluster whose base domain is e.g. `example.com` (`require_subdomain: false`).
2. Dashboard -> Reverse Proxy -> Services -> **Add Service**.
3. In the subdomain field, try to type `dev.app` (to get `dev.app.example.com`). The dot is stripped and you end up with `devapp`.

Compare with the API, which accepts it:

```bash
curl -X POST "https://<management>/api/reverse-proxies/services" \
  -H "Authorization: Token <PAT>" -H "Content-Type: application/json" \
  -d '{
    "name": "dev-app",
    "domain": "dev.app.example.com",
    "mode": "http",
    "enabled": true,
    "targets": [{
      "target_id": "<clusterId>", "target_type": "cluster",
      "protocol": "http", "host": "backend.internal", "port": 80,
      "enabled": true, "options": { "direct_upstream": true }
    }],
    "auth": { "bearer_auth": { "enabled": true, "distribution_groups": [] } }
  }'
# -> 201, meta.status "pending" -> "active"; cert issued for dev.app.example.com;
#    visiting https://dev.app.example.com redirects to the IdP for SSO.
```

### Expected behavior

The subdomain input should allow multi-label subdomains (dot-separated labels), matching what the management API already accepts. Validation should permit a dot-separated sequence of DNS labels (e.g. `^[a-z0-9-]+(\.[a-z0-9-]+)*$`) instead of stripping `.` entirely.

### Actual behavior

Dots are removed on input, so nested subdomains cannot be created from the dashboard; the only workaround is the REST API.

### Are you using NetBird Cloud?

No, self-hosted.

### NetBird version

Management/proxy v0.72.x, dashboard v2.39.0.

### PR DESCRIPTION
**Pull request by khannoussi-malek**
## Issue ticket number and link

Fixes #667 — https://github.com/netbirdio/dashboard/issues/667

The subdomain input in the Reverse Proxy "Add Service" modal stripped dots on
every keystroke, making it impossible to enter multi-label subdomains (e.g.
`dev.app.example.com`) even though the management API already accepts and
provisions them correctly.

### Changes

- New `src/modules/reverse-proxy/domain/subdomain.ts` holding the two concerns
  as separate pure functions, so the live filter and the submit-time check
  can't drift apart:
  - `sanitizeSubdomain` — the live per-keystroke filter, now `[^a-z0-9.-]`.
    Deliberately *not* anchored: users type incrementally, and an anchored
    pattern would reject valid in-progress input like `dev.` on the way to
    `dev.app`.
  - `isValidSubdomain` — anchored `^[a-z0-9-]+(\.[a-z0-9-]+)*$`, so empty
    labels (leading/trailing dot, `dev..app`) still can't reach the API.
- `ReverseProxyDomainInput.tsx` uses `sanitizeSubdomain` in `onChange` and
  surfaces an inline error for malformed input, matching how the existing
  `domainAlreadyExists` error already behaves in this field.
- `ReverseProxyModal.tsx` gates `canContinueToSettings` on `isValidSubdomain`.
  Previously the only subdomain check was `length > 0`, which was sufficient
  only because the input made malformed values unreachable — allowing dots
  removes that guarantee, so the check now has to be explicit.
- `useReverseProxyDomain.ts` uses the same helper when seeding the field from
  a resource name (`initialSubdomain`). This was the same dot-stripping bug on
  a sibling path: a resource named `dev.app` used to seed `devapp`.
- `subdomain.test.ts` covers both functions, following the existing standalone
  script pattern in `src/utils/ip.test.ts` (run with
  `npx tsx src/modules/reverse-proxy/domain/subdomain.test.ts`) — the repo has
  no unit-test runner wired up, so I matched what's already there rather than
  introducing one.

### Testing

- `npx tsx src/modules/reverse-proxy/domain/subdomain.test.ts` — 19 cases pass,
  covering `dev.app` / `a.b.c.d` preserved, and `.app`, `dev.`, `dev..app`,
  `..` rejected.
- `npm run build` — passes.
- Not exercised end-to-end against a live management backend (I don't have one
  running locally), so the modal flow itself is unverified by me. The changed
  logic is pure and covered by the test above; `fullDomain` composition in
  `useReverseProxyDomain` is untouched, so `dev.app` + `example.com` still
  yields `dev.app.example.com`.

Note: `npm run lint` is currently broken on `main` (`next lint` was removed in
Next 16), so I verified formatting with the repo's pinned Prettier and
typechecking via `next build` instead.

## Documentation
Select exactly one:

- [ ] I added/updated documentation for this change
- [x] Documentation is **not needed** for this change (explain why)

This is a bug fix that brings the UI in line with what the management API
already accepts. No documented behavior changes — multi-label subdomains were
already supported server-side, they just couldn't be typed.

### Docs PR URL (required if "docs added" is checked)

n/a

## E2E tests

management-cloud-tag: main
reverse-proxy-tag: main

---
<a href="https://app.blacksmith.sh/netbirdio/codesmith/dashboard/pr/728"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1787439239&installation_model_id=427504&pr_number=728&repository=netbirdio%2Fdashboard&return_to=https%3A%2F%2Fgithub.com%2Fnetbirdio%2Fdashboard%2Fpull%2F728&signature=1315984ecd2e099c87558544afd08f0bc02a3f76deb1e932d4739865ad105ed5"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **New Features**
  * Added shared reverse-proxy subdomain sanitization and validation.
  * The reverse-proxy modal now blocks progression until the subdomain is valid.
  * Subdomain input is normalized as users type, with clear invalid-subdomain messaging.

* **Tests**
  * Added a new TypeScript test suite covering subdomain sanitization and validity rules, including multi-label (dots) and hyphenated forms as well as invalid edge cases.

### CODE REVIEW

File: src/modules/reverse-proxy/domain/subdomain.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// Live input filter. Dots are kept so multi-label subdomains ("dev.app") can
+// be typed at all — dot placement is only checked on submit, because an
+// anchored check would reject valid in-progress input like "dev." mid-typing.
+export function sanitizeSubdomain(value: string): string {
+  return value.toLowerCase().replace(/[^a-z0-9.-]/g, "");
+}
+
+// Dot-separated DNS labels. Empty labels are rejected, so leading/trailing
+// dots and consecutive dots ("dev..app") never reach the API.
+const SUBDOMAIN_PATTERN = /^[a-z0-9-]+(\.[a-z0-9-]+)*$/;
```
[the remark this answers is not in the document]
Comment (khannoussi-malek) [reply src/modules/reverse-proxy/domain/subdomain.ts]: Good catch — fixed in ead8c19.

You are right that the helper was under-delivering on its own contract: I documented it as DNS-label validation but the pattern accepted `-app`, `app-`, and a bare `-`. Applied your regex and added regression cases for leading hyphen, trailing hyphen, bare hyphen, and hyphens at a label boundary in a multi-label value (`dev.-app`, `dev-.app`), plus `a` and `a-b.c-d` on the valid side.

Worth noting this was pre-existing looseness rather than something this PR introduced — the old filter allowed hyphens anywhere and the only check was `length > 0` — but the helper should match what it claims to validate, so it belongs here.

File: src/modules/reverse-proxy/domain/subdomain.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// Live input filter. Dots are kept so multi-label subdomains ("dev.app") can
+// be typed at all — dot placement is only checked on submit, because an
+// anchored check would reject valid in-progress input like "dev." mid-typing.
```
[the remark this answers is not in the document]
Comment (khannoussi-malek) [reply src/modules/reverse-proxy/domain/subdomain.ts]: Fixed in ead8c19 — the comment was stale against my own final code. It now says the sanitizer deliberately skips dot-placement checks (so `dev.` survives mid-typing), and `isValidSubdomain` carries a note that it drives both the inline error and the modal gate.

File: src/modules/reverse-proxy/domain/useReverseProxyDomain.ts [hunk 3]
```diff
@@ -90,11 +91,7 @@ export function useReverseProxyDomain({
   const [subdomain, setSubdomain] = useState(() => {
     return (
       parsed?.subdomain ||
-      initialSubdomain
-        ?.toLowerCase()
-        .replace(/\s+/g, "-")
-        .replace(/[^a-z0-9-]/g, "") ||
-      ""
+      sanitizeSubdomain(initialSubdomain?.replace(/\s+/g, "-") ?? "")
     );
```
[the remark this answers is not in the document]
Comment (khannoussi-malek) [reply src/modules/reverse-proxy/domain/useReverseProxyDomain.ts:95]: Traced this and the bug is real, but I would rather not fix it in this PR.

Confirmed: `domains` comes from `useFetchApi` in `ReverseProxiesProvider` and is passed straight into the modal, which mounts on `modalOpen` with no loading guard. On a cold SWR cache, `parseDomain` runs with `domains === undefined`, falls through to the first-dot split, and `dev.app.example.com` initializes as subdomain `dev` + baseDomain `app.example.com`. The `useState` initializer never re-runs when domains arrive, so it stays wrong.

Two reasons to keep it separate:

1. It is pre-existing on `main`. Both the first-dot fallback and the once-only initializer predate this PR, and nothing here changes that path — this PR only touches the sanitize/validate step. You are right that multi-label domains make it easier to hit, but that is a change in the data, not in the code path.
2. It is the edit flow, not the input-stripping bug in #667, and the fix needs a dirty-ref plus a sync effect — new state machinery that would not be covered by this PR title or its tests.

Happy to open a follow-up issue with the above repro, or send a separate PR if a maintainer would rather have it bundled.
